src/logging.c

```diff
@@ -99,49 +148,96 @@ rcutils_ret_t rcutils_logging_initialize_with_allocator(rcutils_allocator_t allo
-    const char * line_buffered;
+    const char * line_buffered = NULL;
     const char * ret_str = rcutils_get_env("RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED", &line_buffered);
-
     if (NULL == ret_str) {
-      if (strcmp(line_buffered, "1") == 0) {
-        g_force_stdout_line_buffered = true;
-      } else if (strcmp(line_buffered, "0") != 0 && strcmp(line_buffered, "") != 0) {
+      if (strcmp(line_buffered, "") != 0) {
         fprintf(
           stderr,
-          "Warning: unexpected value [%s] specified for RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED. "
-          "Default value 0 will be used. Valid values are 1 or 0.\n",
-          line_buffered);
+          "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED is now ignored.  "
+          "Please set RCUTILS_CONSOLE_USE_STDOUT and RCUTILS_CONSOLE_BUFFERED_STREAM "
+          "to control the stream and the buffering of log messages.\n");
       }
     } else {
       fprintf(
-        stderr, "Error getting env. variable "
+        stderr, "Error getting environment variable "
         "RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s\n", ret_str);
-      ret = RCUTILS_RET_INVALID_ARGUMENT;
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "Error getting environment variable RCUTILS_CONSOLE_STDOUT_LINE_BUFFERED: %s", ret_str);
+      return RCUTILS_RET_ERROR;
+    }
+
+    // Set the default output stream for all severities to stderr so that errors
+    // are propagated immediately.  The user can choose to set the output stream
+    // to stdout by setting the RCUTILS_CONSOLE_USE_STDOUT environment
+    // variable to 1.
+    enum rcutils_get_env_retval retval = rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_USE_STDOUT", "use stderr",
+      "use stdout");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+      case RCUTILS_GET_ENV_ZERO:
+        g_output_stream = stderr;
+        break;
+      case RCUTILS_GET_ENV_ONE:
+        g_output_stream = stdout;
+        break;
     }
 
-    // Check the environment variable for colorized output
-    const char * colorized_output;
-    ret_str = rcutils_get_env("RCUTILS_COLORIZED_OUTPUT", &colorized_output);
+    // Allow the user to choose how buffering on the stream works by setting
+    // RCUTILS_CONSOLE_BUFFERED_STREAM.  With an empty environment variable, use the
+    // default of the stream.  With a value of 0, force the stream to be unbuffered.
+    // With a value of 1, force the stream to be buffered.
+    retval =
+      rcutils_get_env_var_zero_or_one(
+      "RCUTILS_CONSOLE_BUFFERED_STREAM", "not buffered",
+      "buffered");
+    switch (retval) {
+      case RCUTILS_GET_ENV_ERROR:
+        return RCUTILS_RET_INVALID_ARGUMENT;
+      case RCUTILS_GET_ENV_DEFAULT:
+        break;
+      case RCUTILS_GET_ENV_ZERO:
+        if (setvbuf(g_output_stream, NULL, _IONBF, 0) != 0) {
+          char error_string[1024];
```

Comment:
@clalancette meta: nothing against stack allocated char arrays, but I wonder if we should have a statically allocated array that subsequent calls to `rcutils_safe_strerror()` can use, like we do for `rcutils_get_env()`.

Reply:
I think it is a bad idea for threading reasons.  Keeping it stack allocated like this ensures that this can get called from multiple threads.  Granted, this should only be happening at startup time, but I'd still prefer to keep this thread-safe if at all possible.

Alternatives here are to shrink the size of the stack allocated array, or to make it a heap allocation.  I don't really feel strongly about either one (or leaving it as it is).

Reply:
True that `strerror_r` is thread-safe (because `errno` is), unlike the process environment. Disregard.